### DISCUSSION TITLE
Layout Inheritance

### DISCUSSION THREAD

**Discussion by cancan101** [pimcore/pimcore#17033] [Q&A / Support]
Is there any concept of layout inheritance for custom layouts such that I don't have to copy and paste the entire base class definition if I want to modify a small number of fields in the custom layout?

For example if I want to just change the fields to read only, do I need to copy and paste the entire class definition but change `noteditable` to true? or is there a cleaner way to maintain this? Asking because right now keeping all of the layouts in sync is problematic as new fields are added.

**Comment by markus-moser** [comment]
It is not possible to create inherited custom layouts.

If you just would like to change the `readonly` attribute of some fields based on a custom layout you could think about using the `pimcore.admin.dataobject.get.preSendData` event for this to modifiy the JSON which is delivered from the server. 

Here a very fast example which sets the `name` attribute to readonly if the main layout (id '0') is displayed:

```php
<?php
declare(strict_types=1);

namespace App\EventSubscriber;

use Closure;
use Pimcore\Bundle\AdminBundle\Event\AdminEvents;
use Pimcore\Model\DataObject\ClassDefinition;
use Symfony\Component\EventDispatcher\EventSubscriberInterface;
use Symfony\Component\EventDispatcher\GenericEvent;

class PreSendDataSubscriber implements EventSubscriberInterface
{
    public static function getSubscribedEvents()
    {
        return [
            AdminEvents::OBJECT_GET_PRE_SEND_DATA => 'onObjectGetPreSendData',
        ];
    }

    public function onObjectGetPreSendData(GenericEvent $event)
    {
        $data = $event->getArguments()['data'];

       if($data['currentLayoutId'] ?? null == '0') {
            $this->enrichLayoutDefinition($data['layout'], function($node) {
                if($node instanceof ClassDefinition\Data\Input && $node->getName() == 'name') {
                    $node->setNoteditable(true);
                }
            });
       }
    }

    private function enrichLayoutDefinition(ClassDefinition\Data|ClassDefinition\Layout|null &$layout, Closure $modifyNodeFunction): void
    {
        if($layout === null || !method_exists($layout, 'getChildren')) {
            return;
        }

        foreach ($layout->getChildren() ?? [] as $node) {
            $modifyNodeFunction($node);
            $this->enrichLayoutDefinition($node, $modifyNodeFunction);
        }
    }
}
```
